src/types/subst.rs

```diff
@@ -0,0 +1,82 @@
+use std::borrow::Cow;
+
+use crate::extension::ExtensionRegistry;
+
+use super::type_param::{TypeArg, TypeParam};
+use super::{Type, TypeBound, TypeRow};
+
+#[derive(Clone, Debug)]
+pub(crate) struct Substitution<'a> {
+    /// The number of variables bound locally-to the tree within which we are substituting
+    /// (i.e. beneath the point where we started applying the substitution),
+    /// so these should be untouched by the substitution
+    leave_lowest: usize,
+    /// What to do to variables that are affected
+    mapping: Mapping<'a>,
+}
+
+#[derive(Clone, Debug)]
+enum Mapping<'a> {
+    /// An explicit value to substitute for each bound variable
+    Values(Cow<'a, [TypeArg]>),
+    /// An amount to add to the index of any free variable - that is,
+    /// any free var, of index `i`, becomes the variable `(i +` this amount`)`
+    AddToIndex(usize),
+}
+
+impl<'a, T: Into<Cow<'a, [TypeArg]>>> From<T> for Substitution<'a> {
+    fn from(value: T) -> Self {
+        Self {
+            leave_lowest: 0,
+            mapping: Mapping::Values(value.into()),
+        }
+    }
+}
+
+impl<'a> Substitution<'a> {
+    pub(crate) fn apply_var(&self, idx: usize, decl: &TypeParam) -> TypeArg {
+        if idx < self.leave_lowest {
+            return TypeArg::new_var_use(idx, decl.clone());
+        }
+        match &self.mapping {
+            Mapping::Values(args) => args
+                .get(idx - self.leave_lowest)
+                .expect("Unexpected free type var")
+                .clone(),
+            Mapping::AddToIndex(diff) => TypeArg::new_var_use(idx + diff, decl.clone()),
+        }
+    }
+
+    pub(super) fn get_type(&self, idx: usize, bound: TypeBound) -> Type {
+        let TypeArg::Type {ty} = self.apply_var(idx, &TypeParam::Type(bound))
+           else {panic!("Var of kind 'type' did not produce a Type")};
+        ty
+    }
+
+    // A bit unfortunate to need the Extension registry here. We could move `exts` into Substitution?
```

Comment:
Or indeed, move all the `substitute` methods into `Substitution`....(they need some `types`-level access but that's it really)

Reply:
I like the idea of moving all the substitution logic into here 👍

Reply:
Yeah, I think that'd be good to do in a followup. For now I'm trying not to move stuff around unnecessarily just to make the actual change more obvious.

Reply:
I think this would be a good and high priority follow up

Reply:
I've done the small part that's moving the `&ExtensionRegistry` into the Substitution class - that just tidies things up. Moving the rest of the substitute methods can follow later.